node/node.go

```diff
@@ -727,46 +738,78 @@ func (n *Node) OnStop() {
+	// unexported variables.
+
+	// _once is a special object that executes a function only once. We use it to
+	// ensure that `ConfigureRPC` initializes an `Environment` object only once.
+	_once sync.Once
+
+	// _rpcEnv is the `Environment` object serving RPC APIs. We treat it as a
+	// singleton and create it exactly once. See the docs of `ConfigureRPC` below
+	// for more details.
+	_rpcEnv *rpccore.Environment
+)
+
+// ConfigureRPC initializes and returns an `Environment` object with all the data
+// it needs to serve the RPC APIs. The function ensures that the `Environment` is
+// created only once to prevent other parts of the code from creating duplicate
+// `Environment` instances by calling this function directly. This is important
+// because `Environment` stores a copy of the genesis in memory; therefore,
+// multiple independent `Environment` instances would each load the genesis into
+// memory.
 func (n *Node) ConfigureRPC() (*rpccore.Environment, error) {
-	pubKey, err := n.privValidator.GetPubKey()
-	if pubKey == nil || err != nil {
-		return nil, ErrGetPubKey{Err: err}
-	}
-	rpcCoreEnv := rpccore.Environment{
-		ProxyAppQuery:   n.proxyApp.Query(),
-		ProxyAppMempool: n.proxyApp.Mempool(),
+	var errToReturn error
 
-		StateStore:     n.stateStore,
-		BlockStore:     n.blockStore,
-		EvidencePool:   n.evidencePool,
-		ConsensusState: n.consensusState,
-		P2PPeers:       n.sw,
-		P2PTransport:   n,
-		PubKey:         pubKey,
+	_once.Do(func() {
+		pubKey, err := n.privValidator.GetPubKey()
+		if pubKey == nil || err != nil {
+			errToReturn = ErrGetPubKey{Err: err}
+			return
+		}
 
-		GenDoc:           n.genesisDoc,
-		TxIndexer:        n.txIndexer,
-		BlockIndexer:     n.blockIndexer,
-		ConsensusReactor: n.consensusReactor,
-		MempoolReactor:   n.mempoolReactor,
-		EventBus:         n.eventBus,
-		Mempool:          n.mempool,
+		_rpcEnv = &rpccore.Environment{
+			ProxyAppQuery:   n.proxyApp.Query(),
+			ProxyAppMempool: n.proxyApp.Mempool(),
+
+			StateStore:     n.stateStore,
+			BlockStore:     n.blockStore,
+			EvidencePool:   n.evidencePool,
+			ConsensusState: n.consensusState,
+			P2PPeers:       n.sw,
+			P2PTransport:   n,
+			PubKey:         pubKey,
+
+			GenDoc:           n.genesisDoc,
+			TxIndexer:        n.txIndexer,
+			BlockIndexer:     n.blockIndexer,
+			ConsensusReactor: n.consensusReactor,
+			MempoolReactor:   n.mempoolReactor,
+			EventBus:         n.eventBus,
+			Mempool:          n.mempool,
+
+			Logger: n.Logger.With("module", "rpc"),
+
+			Config: *n.config.RPC,
+		}
 
-		Logger: n.Logger.With("module", "rpc"),
+		n.Logger.Info("Creating genesis file chunks if genesis file is too big...")
 
-		Config: *n.config.RPC,
-	}
-	if err := rpcCoreEnv.InitGenesisChunks(); err != nil {
-		return nil, err
-	}
-	return &rpcCoreEnv, nil
+		if err := _rpcEnv.InitGenesisChunks(); err != nil {
+			errToReturn = fmt.Errorf("setting up RPC API environment: %s", err)
```

Comment:
I usually avoid starting error messages with phrases like `failed to do...` or `error...` because it's implicit that an error indicates something went wrong. I think  such phrases are redundant. Another consideration is in Go you chain error strings from nested calls. I find that using `error...` repeatedly can result in awkward and verbose messages. For example, in a call stack of three nested functions, you might get:

```
failed to do X: can't do Y: error doing Z: failed to marshal to JSON
```

Since the innermost error already specifies what actually happened (e.g., failing to marshal to JSON), I believe that focusing on what we were doing when the error occurred makes the message clearer. For example:

```
X: Y: Z: failed to marshal to JSON
```

Here, `X`, `Y`, and `Z` represent the actions being performed at each level, not the function names.

What do you think? Do we have an error convention about this?

Reply:
Hmm... I never thought about it in that way. Makes sense!